### PR TITLE
[fork] monitoring: only send specified alerts

### PR DESCRIPTION
**Pull request by michaellzc**
close https://github.com/sourcegraph/sourcegraph/issues/36320

This PR addes a new config attribute in `observability.alert`, named `alerts`, where user can configure to only receive alerts from the specific list

Builds from this branch can be created with:

```sh
sg ci build docker-images-patch-notest prometheus-gcp
```

https://buildkite.com/sourcegraph/sourcegraph/builds?branch=docker-images-patch-notest/prometheus-gcp/05-31-monitoring_only_send_specificed_alerts

## Test plan

```sh
sg run redis-postgres

sg start monitoring-alerts
```

Configure site-config

> below alerts will always be there on a local environment

```json
  "observability.alerts": [
    {
      "level": "critical",
      "alerts": [
        "warning_frontend_hard_error_search_responses",
        "warning_worker_worker_job_codeintel-janitor_count",
        "critical_worker_worker_job_codeintel-janitor_count"
      ],
      "notifier": {
        "type": "slack",
        "url": "Bring Your Own Webhook (BYOW)"
      }
    }
  ]
```

watch your slack webhook destination 

you can also try a different combination

```json
  "observability.alerts": [
    {
      "level": "critical",
      "alerts": [
        "warning_frontend_hard_error_search_responses",
        "warning_worker_worker_job_codeintel-janitor_count",
        "critical_worker_worker_job_codeintel-janitor_count"
      ],
      "notifier": {
        "type": "slack",
        "url": "Bring Your Own Webhook (BYOW)"
      },
      "owners": ["devops"]
    }
  ]
```

this shouldn't trigger any alerts since it also has a filter on "owners"

You can also visit `sourcegraph.test:9090/alertmanager` to validate the generated config manually

### CODE REVIEW

**Comment by michaellzc** [comment]
Current dependencies on/for this PR:
* main
  * **PR #36321** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/36321" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/36321?utm_source=stack-comment).

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go [hunk 1]
```diff
@@ -145,8 +146,80 @@ For more details, please refer to the service dashboard: %s`, firingBodyTemplate
...
+		if len(alert.Owners) == 0 && len(alert.Alerts) > 0 {
+			// Generate receiver and route for alerts with 'Alerts` but no 'Owners'
+			alerts := strings.Join(alert.Alerts, "|")
+			alertsRegexp, err := amconfig.NewRegexp(fmt.Sprintf("^(%s)$", alerts))
+			if err != nil {
+				newProblem(errors.Errorf("failed to apply alert %d: %w", i, err))
+				continue
+			}
+
+			receiverName := fmt.Sprintf("src-%s-on-%s", alert.Level, randomToken(8))
```
Comment (michaellzc) [inline docker-images/prometheus/cmd/prom-wrapper/receivers.go:158]: there could be many different combination of routing logic, hence just use random token here as receiver name

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go [hunk 2]
```diff
@@ -145,8 +146,80 @@ For more details, please refer to the service dashboard: %s`, firingBodyTemplate
...
+			if r, exists := additionalReceivers[receiverName]; exists {
+				receiver = r
+			} else {
+				receiver = &amconfig.Receiver{Name: receiverName}
+				additionalReceivers[receiverName] = receiver
+				additionalRoutes = append(additionalRoutes, &amconfig.Route{
+					Receiver: receiverName,
+					Match: map[string]string{
```
Comment (daxmc99) [inline docker-images/prometheus/cmd/prom-wrapper/receivers.go:166]: These fields appear to be deprecated and `amconfig` is looking to deprecate them before its `1.0` release. 

Can we use the `matchers` field here instead?

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go [hunk 3]
```diff
@@ -145,8 +146,80 @@ For more details, please refer to the service dashboard: %s`, firingBodyTemplate
...
+						"level": alert.Level,
+					},
+					MatchRE: amconfig.MatchRegexps{
```
Comment (daxmc99) [inline docker-images/prometheus/cmd/prom-wrapper/receivers.go:169]: Same as above, see if we can move to non-deprecated fields

Comment (daxmc99) [reply docker-images/prometheus/cmd/prom-wrapper/receivers.go:169]: We use a fork, this doesn't apply

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go (hunk 2 above)
Comment (daxmc99) [reply docker-images/prometheus/cmd/prom-wrapper/receivers.go:166]: We use a fork, this doesn't apply

**Review by daxmc99 — APPROVED** [review]
@michaellzc 
LGTM 
This likely also needs an entry in the CHANGELOG

**Review by danieldides — APPROVED** [review]
Looks good to me. Were you able to test this locally?

**Comment by michaellzc** [comment]
> Looks good to me. Were you able to test this locally?

yep

the test plan is the instruction to test this locally

File: CHANGELOG.md [hunk 4]
```diff
@@ -21,6 +21,7 @@ All notable changes to Sourcegraph are documented in this file.
 - Extensions: Added site config parameter `extensions.allowOnlySourcegraphAuthoredExtensions`. When enabled only extensions authored by Sourcegraph will be able to be viewed and installed. For more information check out the [docs](https://docs.sourcegraph.com/admin/extensions##allow-only-extensions-authored-by-sourcegraph). [#35054](https://github.com/sourcegraph/sourcegraph/pull/35054)
 - Batch Changes Credentials can now be manually validated. [#35948](https://github.com/sourcegraph/sourcegraph/pull/35948)
 - Zoekt-indexserver has a new debug landing page, `/debug`, which now exposes information about the queue, the list of indexed repositories, and the list of assigned repositories. Admins can reach the debug landing page by selecting Instrumentation > indexed-search-indexer from the site admin view. The debug page is linked at the top. [#346](https://github.com/sourcegraph/zoekt/pull/346)
+- `observability.alerts` now supports a new `alerts` attribute that expects a list of [alerts](https://docs.sourcegraph.com/admin/observability/alert_solutions). When set, you only receive specified alerts. [#36321](https://github.com/sourcegraph/sourcegraph/pull/36321)
```
Comment (bobheadxi) [inline CHANGELOG.md:24]:
```suggestion
- `observability.alerts` now supports a new `alerts` attribute that expects a list of [alert identifiers](https://docs.sourcegraph.com/admin/observability/alert_solutions) (the same as those used for `silenceAlerts`). When set, you only receive specified alerts. [#36321](https://github.com/sourcegraph/sourcegraph/pull/36321)
```

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go [hunk 5]
```diff
@@ -145,8 +146,80 @@ For more details, please refer to the service dashboard: %s`, firingBodyTemplate
...
+			}
+
+			alerts := strings.Join(alert.Alerts, "|")
+			alertsRegexp, err := amconfig.NewRegexp(fmt.Sprintf("^(%s)$", alerts))
+			if err != nil {
+				newProblem(errors.Errorf("failed to apply alert %d: %w", i, err))
+				continue
+			}
+
+			receiverName := fmt.Sprintf("src-%s-on-%s", alert.Level, owners)
```
Comment (bobheadxi) [inline docker-images/prometheus/cmd/prom-wrapper/receivers.go:197]: Is this receiver name still accurate?

File: docker-images/prometheus/cmd/prom-wrapper/receivers.go (hunk 1 above)
Comment (bobheadxi) [reply docker-images/prometheus/cmd/prom-wrapper/receivers.go:158]: I've actually forgotten the reason behind the receiver name, let's review this tomorrow 😆

File: schema/site.schema.json [hunk 6]
```diff
@@ -1191,6 +1191,13 @@
             "items": {
               "type": "string"
             }
+          },
+          "alerts": {
+            "description": "Do not use. When set, only receive specified alerts. Used by Sourcegraph internally.",
```
Comment (bobheadxi) [inline schema/site.schema.json:1196]: If we say `Do not use`, should we _not_ advertise it in the changelog?

**Review by bobheadxi — COMMENTED** [review]
very cool 😁 will do a more in-depth review tomorrow

**Comment by michaellzc** [comment]
> Are we comfortable with assuming that our opt-in list will remain up-to-date at all times? If we add a new alert that should be enabled on MIs, we need to have a process in place that ensures we update all instances accordingly.
> 
> An alternative approach would be the inverse of this implementation, making it opt-out instead. That would mean a worst-case scenario is we get alerted too much rather than too little.
> 
> If we stick to this implementation, let's think about how we keep the inclusion list up-to-date.

we will still be notified for alerts that not not whitelisted, but in a different channel (#alerts-managed-instances)

the whitelist feature is ensuring we only route things we care about the most and require immediate attention to Opsgenie.

**Comment by bobheadxi** [comment]
Rebased and rebuilding: https://buildkite.com/sourcegraph/sourcegraph/builds/173571

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| -0.00% (-0.02 kb) | 0.00% (+0.13 kb) | 0.00% (+0.15 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/05-31-monitoring_only_send_specificed_alerts/compare-report.html;tab=live_object) for a full comparison between the commits 7d0178c and b8da938 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

**Pull request closed without merging**